terminus/builders/osm_city_builder.py

```diff
@@ -0,0 +1,122 @@
+                                (self.bounds['minlon'] + self.bounds['maxlon']) / 2,
+                                0)
+
+        logger.debug("Map Origin: {0}".format(self.map_origin))
+
+        # Calculate bounding box based on lat/lon
+        min_xy = self._translate_coords(self.bounds['minlat'],
+                                        self.bounds['minlon'])
+        max_xy = self._translate_coords(self.bounds['maxlat'],
+                                        self.bounds['maxlon'])
+        self.bounds['min_xy'] = min_xy
+        self.bounds['max_xy'] = max_xy
+
+        self.bounds['size'] = Point(abs(min_xy.x) + abs(max_xy.x),
+                                    abs(min_xy.y) + abs(max_xy.y),
+                                    0)
+
+        logger.debug("Bound box size {0}".format(self.bounds['size']))
+        logger.debug(
+            "Bound box 'minlat': {0}, 'maxlat': {1}\n\t\t\t\t\t  "
+            "'minlon': {2}, 'maxlon': {3}".format(self.bounds['minlat'],
+                                                  self.bounds['maxlat'],
+                                                  self.bounds['minlon'],
+                                                  self.bounds['maxlon'])
+        )
+
+        # Parse OSM map
+        self.parser = OSMParser(coords_callback=self._get_coords,
+                                ways_callback=self._get_ways)
+        self.parser.parse(self.osm_map)
+
+        logger.debug("Number of ways: {}".format(len(self.osm_ways)))
+        logger.debug("Number of coords: {}".format(len(self.osm_coords)))
+
+        # Ground plane
+        city.set_ground_plane(GroundPlane(max(self.bounds['size'].x,
+                                              self.bounds['size'].y),
+                                          Point(0, 0, 0)))
+
+        self._create_roads(city)
+
+        return city
+
+    def _get_ways(self, ways):
+        ''' OSM parser callback for the ways '''
+        # For now exclude smaller roads, in the future we should draw each
+        # type of street with different width/texture
+        road_types = ['motorway', 'trunk', 'primary', 'secondary', 'tertiary',
+                      'unclassified', 'residential']
+
+        for osmid, tags, refs in ways:
+            if 'highway' in tags:
+                if tags['highway'] in road_types:
+                    self.osm_ways[osmid] = {'tags': tags, 'refs': refs}
+
+    def _get_coords(self, coords):
+        ''' OSM parser callback for the coords '''
+        for osmid, lon, lat in coords:
+            self.osm_coords[osmid] = {'lat': lat, 'lon': lon}
+
+    def _create_roads(self, city):
+        for key, value in self.osm_ways.iteritems():
+            tmp_road = Street()
+            for ref in value['refs']:
+                # Check if coord is inside bounding box
+                if self.bounds['minlat'] < self.osm_coords[ref]['lat'] < self.bounds['maxlat'] and \
```

Comment:
hmm, I think we may have an issue here if the road starts inside the bounding box, goes out and then in again (think e.g. of a roundabout). I think that if we only trim the edges of the street we should be ok, but we have to avoid removing intermediate points, otherwise we will get a fake geometry.